bqskit/passes/processing/treescan.py

```diff
@@ -0,0 +1,190 @@
+            raise TypeError(
+                'Expected cost to be a CostFunctionGenerator, got %s'
+                % type(cost),
+            )
+
+        if not isinstance(instantiate_options, dict):
+            raise TypeError(
+                'Expected dictionary for instantiate_options, got %s.'
+                % type(instantiate_options),
+            )
+
+        self.collection_filter = collection_filter or default_collection_filter
+
+        if not callable(self.collection_filter):
+            raise TypeError(
+                'Expected callable method that maps Operations to booleans for'
+                ' collection_filter, got %s.' % type(self.collection_filter),
+            )
+
+        self.tree_depth = tree_depth
+        self.start_from_left = start_from_left
+        self.success_threshold = success_threshold
+        self.cost = cost
+        self.instantiate_options: dict[str, Any] = {
+            'dist_tol': self.success_threshold,
+            'min_iters': 100,
+            'cost_fn_gen': self.cost,
+        }
+        self.instantiate_options.update(instantiate_options)
+
+    @staticmethod
+    def get_tree_circs(
+        orig_num_cycles: int, circuit_copy: Circuit,
+        cycle_and_ops: list[tuple[int, Operation]],
+    ) -> list[Circuit]:
+        # Implement recursively for now, if slow then fix
+        all_circs = [circuit_copy.copy()]
+        for cycle, op in cycle_and_ops:
+            new_circs = []
+            for circ in all_circs:
+                idx_shift = orig_num_cycles - circ.num_cycles
+                new_cycle = cycle - idx_shift
+                work_copy = circ.copy()
+                work_copy.pop((new_cycle, op.location[0]))
+                new_circs.append(work_copy)
+                new_circs.append(circ)
+
+            all_circs = new_circs
+
+        all_circs = sorted(all_circs, key=lambda x: x.num_operations)
+
+        return all_circs
+
+    async def run(self, circuit: Circuit, data: PassData) -> None:
+        """Perform the pass's operation, see :class:`BasePass` for more."""
+        instantiate_options = self.instantiate_options.copy()
+        if 'seed' not in instantiate_options:
+            instantiate_options['seed'] = data.seed
+
+        start = 'left' if self.start_from_left else 'right'
+        _logger.debug(f'Starting scanning gate removal on the {start}.')
+
+        target = self.get_target(circuit, data)
+        # target = None
+
+        circuit_copy = circuit.copy()
+        reverse_iter = not self.start_from_left
+
+        ops_left = list(circuit.operations_with_cycles(reverse=reverse_iter))
+        print(
+            f'Starting Scan with tree depth {self.tree_depth}'
+            ' on circuit with {len(ops_left)} gates',
+        )
+
+        while ops_left:
+            chunk = ops_left[:self.tree_depth]
+            ops_left = ops_left[self.tree_depth:]
+
+            # Circuits of size 2 ** tree_depth - 1,
```

Comment:
"Circuits of size" is a bit misleading, it might be better to say "2 ** tree_depth - 1 Circuits".

Moreover, this comment would be redundant with a proper docstring for `get_tree_circs`